### PR TITLE
The cross compiled unit tests for STM32 and AVR now compile again.

### PR DESCRIPTION
**Pull request by strongly-typed**
The STM32 unit test was run on an STM32F4Discovery board and revealed some minor problems. 
The unit test for AVR was not yet run in real hardware. 

This is WiP.

Even though the unit tests are not run on the actual hardware yet, it is required that the tests compile before someone actually can take care about actually **running** these tests. So please let TravisCI build the cross unit test from now on and take care they stay compilable.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -16,6 +16,8 @@ Remember that you can continue to push to your `feature/*` or `fix/*` branches t
 
 Please note that your pull request will have to pass the CI tests in order to be merged.
 
+Including unit tests is very welcomed and highly recommended. xpcc has its own simple and easy to use unit test framework. Unit tests are placed in the `test` subdirectory of every component (e.g. `xpcc/src/xpcc/math/filter/test`) and are run with `scons unittest`. See TESTING.md for more information about testing strategies in xpcc.
+
```
Comment (salkinium) [inline CONTRIBUTING.md:20]: :+1: :fireworks: :tada:

File: templates/unittest/runner_atmega.cpp.in [hunk 2]
```diff
@@ -7,42 +7,49 @@
...
 	LedRed::set();
-	
+
 	// initialize output device
-	xpcc::atmega::BufferedUart0 uart(115200);
-	xpcc::IODeviceWrapper<xpcc::atmega::BufferedUart0> device(uart);
-	
+	GpioOutputE1::connect(Uart0::Tx);
+    GpioInputE0::connect(Uart0::Rx);
+    Uart0::initialize< clock, 115200, xpcc::Tolerance::TenPercent >();
```
Comment (salkinium) [inline templates/unittest/runner_atmega.cpp.in]: Indentation. _ducks_ :stuck_out_tongue_winking_eye:

File: CONTRIBUTING.md (hunk 1 above)
Comment (salkinium) [inline CONTRIBUTING.md:20]: This is too good to not put it into the README.md TL;DR section at the top!
Maybe something like this?

``` md
[…]
- We have Continuous Integration as well.
- [And we care a lot about testing](TESTING.md).
- You have questions?
[…]
```

Comment (salkinium) [inline CONTRIBUTING.md:20]: Also, in that text up there, link to the files (ie. TESTING.md and source folder), so that people don't have to search for it.
If they looking at this file, they will probably already be on the website.

**Comment by strongly-typed** [comment]
Will fix this in my rail office today ;-)

**Comment by strongly-typed** [comment]
Do you know why there is no TravisCI build for this PR?

File: TESTING.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	$ scons unittest
+
+## Unit Tests on Target Platform
+
+A very specific feature of the xpcc unit test framework is that the unit tests can be run on the target platform. This matters because in most cases xpcc is used for cross compiling and the target platform differs at least in one of the following features
+
+ * compiler (gcc for x86 vs. arm-none-eabi vs. avr-gcc)
+ * word size (64 bits vs. 8 or 32 bits)
+ * endianess
+ * presence of and FPU (i7 vs. AVR or ARM Cortex-4Mf)
```
Comment (salkinium) [inline TESTING.md]: s/and/a

**Comment by strongly-typed** [comment]
@salkinium Could you please trigger TravisCI?
Rewriting history does not trigger a TravisCI build.

File: README.md [hunk 4]
```diff
@@ -13,10 +13,12 @@ which is efficient enough to be deployed on a small ATtiny, yet powerful enough
 use of advanced capabilities found on the 32bit ARM Cortex-M.
 
 TL;DR:
+
 - [This project has a homepage](http://xpcc.io) with [install instructions](http://xpcc.io/install.html) and [a technical blog](http://blog.xpcc.io).
 - [Feast your eyes on lots of working examples](examples).
 - [API documentation is available too](http://xpcc.io/api/modules.html).
 - [We have Continuous Integration as well](https://travis-ci.org/roboterclubaachen/xpcc).
+- [We care a lot about testing and are keen to develop powerful testing strategies for embedded devices.](TESTING.md)
```
Comment (salkinium) [inline README.md]: Should this be shorter? It seems a little… "markety" compared to the others.

**Comment by salkinium** [comment]
Could you squash "Clear buffer of MemoryWriter when clear() is called." together?

> Rewriting history does not trigger a TravisCI build.

In my experience it is a bit fidgety sometimes :-(

**Comment by salkinium** [comment]
ping @strongly-typed

**Comment by strongly-typed** [comment]
Tried to catchup with develop.

@ekiwi @salkinium The linkescript is not generated for target=stm32

See
`https://travis-ci.org/roboterclubaachen/xpcc/jobs/127037134`

Copying the script manually works

`cp ./build/release/test/stm32f407vg/libxpcc/generated_platform/driver/core/cortex/linkerscript.ld build/unittest_stm32/libxpcc/generated_platform/driver/core/cortex`

**Comment by salkinium** [comment]
Yes, see also my failed integration: https://github.com/salkinium/xpcc/commit/7d84b1ec604f5888d5e61e725590ca075e2313d7

**Comment by salkinium** [comment]
See the hacky solution here, which works and is good enough for now:
https://github.com/salkinium/xpcc/commit/f5122481aa499bc9ce322d3fe28720cb9634e814

Would you like to take my branch of this? Everything is the same, except it is rebased on the latest changes and contains two commits, which integrate this a little better with the changes to the `project.cfg` regarding dev boards.

**Comment by strongly-typed** [comment]
@salkinium Cherry-picking your commit seems to fix this.

**Comment by salkinium** [comment]
I've merged this manually, since I fixed the `build.board` setting in `project.cfg`.

I wish GitHub had a "Merged Manually" button, so i didn't have to close this. Hmpf.

**Comment by salkinium** [comment]
PS: Your testing guide is now also available directly on the website: http://xpcc.io/guide/testing/
